STM32F1/platform.txt

```diff
@@ -45,7 +45,7 @@ build.cpu_flags=
 build.hs_flag=
 build.upload_flags=
 build.flags.optimize=-Os
-build.flags.ldspecs=
+build.flags.ldspecs=--specs=nano.specs

```

Comment:
I have the feeling that this option will increase the code site instead of reducing it.
I downloaded the STM32DUINO package in Arduino boards manager ( which took 1 GB space !!!) and changed the platform.txt to use the xpack instead of old EABI package.
Without nano-specs the code size was 14,704 bytes, with it was 23,556 bytes....

So am have huge doubts whether this PR really brings any advantage to this core.

Reply:
> I have the feeling that this option will increase the code size instead of reducing it.

No. If this is the case then you have a mistake with the config.

> STM32DUINO package

when you told this you talk about which one ?

Reply:
I mean STM32DUINO package the official ST package.

Regarding the code size, what kind of configuration do you mean? I just tested the platform.txt by adding the mentioned flag, and the outcome was as I wrote, an increase in size.

Reply:
OK, I misunderstood your point. You download it to get the gcc and try with your core 😉 

Anyway, I've already test in the past and did not get those results. Maybe some libc functions used by the core are huge with the newlib nano. For example assert.h usage is huge. Also usage of float for printf and scanf can increase size anyway I'm really surprise of your result as by default this should be less.

Reply:
Here my test with the master of the Arduino_STM432 repo:

# Blink example

### Default config: gcc 7-2017q4 and no nanospecs
Sketch uses 15836 bytes (14%) of program storage space. Maximum is 110592 bytes.
Global variables use 3176 bytes (18%) of dynamic memory, leaving 14232 bytes for local variables. Maximum is 17408 bytes.

### Config: gcc 7-2017q4 and nanospecs
Sketch uses 14276 bytes (12%) of program storage space. Maximum is 110592 bytes.
Global variables use 2160 bytes (12%) of dynamic memory, leaving 15248 bytes for local variables. Maximum is 17408 bytes.

### Config: xpaxk gcc 9.2.1-1.1 and no nanospecs
Sketch uses 15736 bytes (14%) of program storage space. Maximum is 110592 bytes.
Global variables use 3176 bytes (18%) of dynamic memory, leaving 14232 bytes for local variables. Maximum is 17408 bytes.

### Config: xpaxk gcc 9.2.1-1.1 and nanospecs
Sketch uses 14272 bytes (12%) of program storage space. Maximum is 110592 bytes.
Global variables use 2160 bytes (12%) of dynamic memory, leaving 15248 bytes for local variables. Maximum is 17408 bytes.

So there is a benefits with nanospecs but with the toolchain it is not so evident. But I guess with a real sketch using more feature it would be better.

Anyway, I really don't understand how you get your results.

Reply:
OK, I have to correct my information, previously I did the test with my repo, now using Roger's repo:

arm-none-eabi-gcc:
```
Sketch uses 15384 bytes (23%) of program storage space. Maximum is 65536 bytes.
Global variables use 3176 bytes (15%) of dynamic memory, leaving 17304 bytes for local variables. Maximum is 20480 bytes.
```

xpack-arm-none-eabi-gcc-9.2.1-1.1:
```
Sketch uses 15316 bytes (23%) of program storage space. Maximum is 65536 bytes.
Global variables use 3176 bytes (15%) of dynamic memory, leaving 17304 bytes for local variables. Maximum is 20480 bytes.
```

arm-none-eabi-gcc + nano.specs:
```
Sketch uses 13940 bytes (21%) of program storage space. Maximum is 65536 bytes.
Global variables use 2064 bytes (10%) of dynamic memory, leaving 18416 bytes for local variables. Maximum is 20480 bytes.
```

xpack-arm-none-eabi-gcc-9.2.1-1.1 +nano.specs:
```
Sketch uses 13936 bytes (21%) of program storage space. Maximum is 65536 bytes.
Global variables use 2064 bytes (10%) of dynamic memory, leaving 18416 bytes for local variables. Maximum is 20480 bytes.
```

As a conclusion, while using the nano,specs brings visible code size improvement, it does not bring to much additionally when using the xpack-arm-none-eabi-gcc-9.2.1-1.1.

So at this point I am still not very motivated to load ~1GB of data (the official core including XPACK) to reduce only a few bytes in the code size.

Reply:
I agree with Steve, the space saving by using xpack-arm-none-eabi-gcc-9.2.1-1.1. is minimal, and there would need to be a huge amount of manual regression testing to confirm that xpack-arm-none-eabi-gcc-9.2.1-1.1. does not cause any runtime problems with code, including things like RTOS compatibility, and IRQ, DMR , USB compatibility etc

Re:Nanospecs
I don't know the details of what library code gets removed when Nanospecs is used, normally you can't reduce the binary size by 10%+ with loosing some functionality.
The core may not be using the functionality, but some existing application code may be.

Reply:
simply add the newlib nano specs as an option and kept standard newlib as default.